grpc-kotlin/src/main/kotlin/com/linecorp/armeria/server/grpc/kotlin/CoroutineServerInterceptor.kt

```diff
@@ -0,0 +1,85 @@
+package com.linecorp.armeria.server.grpc.kotlin
+
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import com.linecorp.armeria.internal.common.kotlin.ArmeriaRequestCoroutineContext
+import com.linecorp.armeria.internal.server.grpc.AbstractServerCall
+import com.linecorp.armeria.server.grpc.AsyncServerInterceptor
+import io.grpc.Metadata
+import io.grpc.ServerCall
+import io.grpc.ServerCallHandler
+import io.grpc.ServerInterceptor
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.future.future
+import java.util.concurrent.CompletableFuture
+
+/**
+ * A [ServerInterceptor] that is able to asynchronously execute the interceptor without blocking the
+ * caller thread.
+ * For example:
+ * ```kotlin
+ * class AuthInterceptor : CoroutineServerInterceptor {
+ *     override suspend fun <ReqT, RespT> suspendedInterceptCall(
+ *             call: ServerCall<ReqT, RespT>,
+ *             headers: Metadata,
+ *             next: ServerCallHandler<ReqT, RespT>
+ *     ): ServerCall.Listener<ReqT> {
+ *         val result = authorizer.authorize(ServiceRequestContext.current(), headers).await()
+ *         if (result) {
+ *             return next.startCall(call, headers)
+ *         } else {
+ *             throw AuthenticationException("Invalid access")
+ *         }
+ *     }
+ * }
+ * ```
+ */
+interface CoroutineServerInterceptor : AsyncServerInterceptor {
```

Comment:
```suggestion
@UnstableApi
interface CoroutineServerInterceptor : AsyncServerInterceptor {
```

Reply:
Done in [63e0b79](https://github.com/line/armeria/pull/4724/commits/63e0b79cdf01d14c58d916c29c87d1f129f255a8)!

Reply:
@ikhoon Should we use `@RequiresOptIn` for Kotlin? 🤔

Reply:
`@RequiresOptIn` sounds good. Users will easily know what API is unstable from it.

Reply:
<img width="298" alt="image" src="https://user-images.githubusercontent.com/37795866/224260233-ae92c78d-9576-41e9-bcfe-68990db709f3.png">

Unfortunately, We can't use `@RequiresOptIn` to interface 😭

Reply:
@trustin said `@RequiresOptIn` seems to be available indirectly in other annotation classes.

So it would be good to add the Kotlin version of `@UnstableApi` and use this. 
It would be nice to have it separate from this PR 👍